kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/InternalCompletableFuture.java

```diff
@@ -10,20 +10,28 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import io.netty.util.concurrent.ThreadAwareExecutor;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
 
 /**
- * Implementation of CompletableFuture that uses the given {@link Executor}
- * for async method invocations that do not specify one.
+ * Implementation of CompletableFuture that executes all chained work on a specific
+ * event loop.
```

Comment:
The golden rule when working with Netty is don't block the event loop! The problem I've just realised with this and all the other `Async()` signatures is we are changing the expected semantics as we override the defaultExecutor for the future to be the eventloop. 

The use case in question would be: 

```
context.sendRequest().thenComposeAsync(message -> Files.readAllLines("/path/to/HugeFile.csv").forEach(s -> s.contains(message.apiKey())));
```

Contrived example for the sake of making a point.

The expected semantic is that calling `Async` has moved the work from this thread to the background (generally the Common fork join pool). However thats not what is happening when we return an ICF.

Reply:
yep worth moving to a discussion. Our original choice documented on `sendRequest` was

```
<p>Default and asynchronous default computation stages chained to the returned
     * {@link java.util.concurrent.CompletionStage} are guaranteed to be executed by the thread associated with the
     * connection.
```

Which I think was more about safety, enabling users to mutate the filter state from *Async methods. But as you said it does make the *Async family surprising and unable to be used to shunt work off the eventloop, users have to use the variant where they supply their own executor.